Render interface

ivovandongen opened the pull request:
Introduces a `Renderer` interface with a matching `RendererFrontend` interface to decouple the map from the actual rendering. This makes it possible to have different threading strategies per platform.

ivovandongen commented:
@jfirebaugh Given your notes in https://github.com/mapbox/mapbox-gl-native/issues/8820 I've started implementing a renderer abstraction and proxy/bridge interface for the SDK to implement so we can do either synchronous or asynchronous rendering. 

As I understood your notes, we want something like:

![render-interface](https://user-images.githubusercontent.com/64527/27399099-88109d74-5670-11e7-9487-3033e5ab1286.png)

Besides quite some details to work out I'm running into a issue with the sdk binding part you might have already thought about:

The proxy/bridge ideally doesn't know about the data passed to the renderer as it contains references to non-public headers (like the TransformState). I can find no way to pass it along with unique_ptrs though, without having the definition been known at the point where I move the data. 
Atm it works when the renderer accepts const references (https://github.com/mapbox/mapbox-gl-native/pull/9327/commits/9f4c3ec1842a301e4fcef08b077abf7640668afe#diff-9f4c7224e9bcf6a1be79b87e1591357dR47). But when using different threads we want to actually move the data and accept a unique_ptr in the renderer as well so that the data is not destroyed before it has been used in the renderer. This invariably leads to `error: invalid application of 'sizeof' to an incomplete type`. A way around this probably would be to use an Impl for each of the messages, but that will add quite a lot of code.

jfirebaugh commented:
Current call chain for continuous rendering is:

1. style mutation
2. `onUpdate(Update::Repaint)`
3. `asyncInvalidate.send()`
4. ..........
5. `backend.invalidate()`
6. <SDK code>
7. `Map::render(view)`
8. `renderStyle->update`, `painter->render`

With the introduction of `RenderBridge`, we should reduce the number of moving parts here.

* Replace `onUpdate(Update::Repaint)` with `renderBridge.update(...)`
* Remove `Backend::invalidate`
* Remove `Map::Impl::asyncInvalidate`; instead `RenderBridge` implementation is responsible for throttling repaints
* Merge `View` into `Backend`
* `RenderBridge` must provide `Backend&` to `Renderer::render`
* In `Map` constructor parameters, replace `Backend&` with `RenderBridge&`; remove `createRenderBridge()`

This will make a clean division:

* `RenderBridge` is the interface consumed by `Map` on the main thread.
* `Backend` is the interface consumed by `Renderer` on the orchestration/rendering thread.

ivovandongen commented:
@jfirebaugh I made some progress with reducing the number of moving parts. Leaving some notes here.

- [x] Replace onUpdate(Update::Repaint) with renderBridge.update(...)
As discussed, we need to measure if the additional creation of `UpdateParameters` impacts performance. I fixed up the node and mac os bindings. Render tests succeed, but the rendering on the macos app stutters. Need to investigate if this is the cause.

- [x] Remove Backend::invalidate
Removed it from Backend, but kept it intact in the way that a invalidate callback can be provided to the default render bridge so the UI system can be used to synchronise render calls with rendering the rest of the UI like before

- [x] Remove Map::Impl::asyncInvalidate; instead RenderBridge implementation is responsible for coalescing updates/repaints
I've implemented this in the `DefaultRenderBridge` since it is common to a couple of our platforms and I don't want to duplicate the code. ATM it provides either a constructor that takes a callback or the `Backend` and `View` so that `Renderer::render` may be called after coalescing updates

- [x] RenderBridge must provide Backend& to Renderer::render
Done through the invalidate callback / constructor arguments for now. Maybe inheritance is more suitable here. 

- [x] In Map constructor parameters, replace Backend& with RenderBridge&; remove createRenderBridge()
We probably want to split off the MapObserver from the `RenderBridge&` argument so that we more easily use composition in the SDK bindings instead of inheritance. ATM the SDK view implementations are a combination of Backend, RenderBridge, View and MapObserver. Which makes re-use of, `DefaultRenderBridge` for example a bit clunky. For example, in the macos binding: `_mbglView->render(*_mbglView, *_mbglView);`

related: Cutting the ownership between `Map` and `RenderStyle` and providing it indirectly by reference to the `RenderBridge` led me to a bit of an awkward position. As the `RenderStyle` needs to be destructed before the `AnnotationManager` since the former uses the latter through the destructor of `AnnotationTile` we need to ensure that the Renderer/RenderBridge gets destructed before the Map, but since the Map needs the RenderBridge as a reference now we can't declare both on the stack anymore (because we need to explicitly delete them in the same order). 
For example, the "fixed" Annotations test:

```
    std::unique_ptr<RenderBridge> renderBridge;
    std::unique_ptr<Map> map;

    AnnotationTest()
            : renderBridge(std::make_unique<DefaultRenderBridge>(backend, view))
            , map(std::make_unique<Map>(*renderBridge, view.getSize(), 1, fileSource, threadPool, MapMode::Still)) {

    }

    ~AnnotationTest() {
        renderBridge.reset();
        map.reset();
    }
```

When we use the renderer on a separate thread, we are going to have to assure that the destruction of the Renderer/RenderBridge still completes before destructing the map.

- Merge View into Backend
Started on this, but don't know if this makes sense for all implementations, for example `OffscreenTexture` and `OffscreenView`. The View is often re-created as well in the node bindings whereas the Backend and associated Context can remain intact.

jfirebaugh commented:
> Removed it from Backend, but kept it intact in the way that a invalidate callback can be provided to the default render bridge so the UI system can be used to synchronise render calls with rendering the rest of the UI like before

I think instead we should have per-SDK implementations of `RenderBackend`, rather than trying to share code with `DefaultRenderBackend`.

> I've implemented this in the `DefaultRenderBridge` since it is common to a couple of our platforms and I don't want to duplicate the code.

It's not that much duplication, and some SDKs probably don't even need it. E.g. `[nativeView setNeedsGLDisplay]` likely already does coalescing internally.

> We probably want to split off the `MapObserver` from the `RenderBridge&` argument

Yes, good idea.

> Cutting the ownership between `Map` and `RenderStyle` and providing it indirectly by reference to the `RenderBridge` led me to a bit of an awkward position.

Does this go away if ownership of `RenderBridge` is passed into `Map`, i.e. the parameter is `std::unique_ptr<RenderBridge>` rather than `RenderBridge&`, and `Map` releases it explicitly in `~Map`?

jfirebaugh commented on src/mbgl/map/map.cpp:
```diff
@@ -820,36 +759,32 @@ MapDebugOptions Map::getDebug() const {
-        impl->renderStyle->setSourceTileCacheSize(size);
-        impl->backend.invalidate();
-    }
+    impl->backend.setRendererState([=](Renderer* renderer) {
```
Let's fix #8039 now so that we don't need this.

ivovandongen replied:
Good idea. I'll start on that first.

jfirebaugh commented on src/mbgl/map/map.cpp:
```diff
@@ -820,36 +759,32 @@ MapDebugOptions Map::getDebug() const {
-        impl->renderStyle->onLowMemory();
-        impl->backend.invalidate();
-    }
+    impl->backend.setRendererState([](Renderer* renderer) {
```
Let's have an explicit pass-through in `RenderBridge` for this.

ivovandongen replied:
👍

jfirebaugh commented on src/mbgl/map/map.cpp:
```diff
@@ -888,9 +823,11 @@ void Map::Impl::onResourceError(std::exception_ptr error) {
-    if (impl->renderStyle) {
-        impl->renderStyle->dumpDebugLogs();
-    }
+    impl->backend.setRendererState([](Renderer* renderer) {
```
And this. And then remove `setRendererState`.

ivovandongen commented:
> I think instead we should have per-SDK implementations of RenderBackend, rather than trying to share code with DefaultRenderBackend.

Makes sense if coallescing is indeed a platform issue. I'll try it out on a couple of platforms. I know that Android also just sets a flag that isn't used until the next draw call.

>> Cutting the ownership between Map and RenderStyle and providing it indirectly by reference to the RenderBridge led me to a bit of an awkward position.

> Does this go away if ownership of RenderBridge is passed into Map, i.e. the parameter is std::unique_ptr<RenderBridge> rather than RenderBridge&, and Map releases it explicitly in ~Map?

That would help a lot. I used a reference initially as moving ownership limits the use of the RenderBridge as a part of the platform View bindings by inheritance (`MGLMapView` for example). But the multiple-inheritance there doesn't make for a very clear abstraction anyway.

ivovandongen commented:
@jfirebaugh I'm finishing up with the platform changes. Besides a bug specific to egl on linux, they seem to work well. Could you give it a quick check if it's moving in the right direction?

Major parts still on my todo list:
- Rename/move `Backend`
- Rename/move `BackendScope`
- Get rid of the `renderer_impl.hpp` import in `map.cpp` (move out some of the parameters definitions)
- `Map::isFullyLoaded`
- `Map::triggerRepaint()`

For reference the bug with egl/linux

```
[----------] 1 test from AnnotationTile
[ RUN      ] AnnotationTile.Issue8289
terminate called after throwing an instance of 'std::runtime_error'
  what():  Error creating the EGL context object.
```

ivovandongen commented:
@jfirebaugh Noting additional changes from our conversation on chat:

- [ ] Extract the removal of `MapObserver` from `Backend` in a separate commit
- [ ] `Renderer`:
    - [ ] Let SDK's create the `Renderer`
    - [ ] Remove redundant parameters from `Map` constructor
    - [ ] Move `BackendScope` handing into the renderer (Obtain `BackendScope` from `Backend` / or at least the `ScopeType`)
    - [ ] Move `Backend` into constructor parameters as it can't be switched during the life-time of `Renderer` anyway
    - [ ] Get rid of `TransformState` from query parameters and re-use the one in the `Renderer`. This allows us to get rid of the *QueryParameters wrappers as well.
    - [ ] (maybe) Combine query* methods by using a co-variant argument
- `RendererFrontend`
    - [ ] Add `resetRenderer()` method to force a synchronous cleanup of the internals (to deal with the fact that `AnnotationManager` is shared between `Map` and `RenderStyle` ATM)
    - [ ] Go back to passing the `RendererFrontend` by reference to the `Map`. This enables us to implement the `RendererFrontend` interface as part of other (UI) objects on the SDK side.
    - [ ] Remove `onLowMemory()` and instead call directly from SDK bindings to `Renderer` where needed 
    - [ ] Remove `dumpDebugLogs()` and instead call directly from SDK bindings (node) to `Renderer` where needed 
- [ ] `AnnotationManager`
    - [ ] Remove reference to style and deal with `Map::setStyle`
- [ ] `Map`
    - [ ] `Map::isFullyLoaded()` Reuse and store state from `RendererObserver::onDidFinishRenderingFrame`.
- [ ] General
    - [ ] Clean up unused imports
    - [ ] Clean up comments

Plus the outstanding work for completeness:
- [ ] Rename/move Backend
- [ ] Rename/move BackendScope
- [ ] Get rid of the zrenderer_impl.hpp` import in `map.cpp`
- [ ] Fix egl+linux builds

ivovandongen commented:
I fixed the EGL errors. It appears that the `EGLDisplay` caching in `test::sharedDisplay()` was not working as expected. At certain points the `EGLDisplay::Impl` would be destructed, causing `eglTerminate()` to be called. However, there would not be a subsequent call to `eglIntialize()` from as no new  `EGLDisplay` was created. I removed the display caching all-together.

frederoni commented on platform/ios/ios.xcodeproj/project.pbxproj:
```diff
@@ -212,6 +212,8 @@
 		7E016D871D9E890300A29A21 /* MGLPolygon+MGLAdditions.m in Sources */ = {isa = PBXBuildFile; fileRef = 7E016D831D9E890300A29A21 /* MGLPolygon+MGLAdditions.m */; };
 		920A3E5D1E6F995200C16EFC /* MGLSourceQueryTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 920A3E5C1E6F995200C16EFC /* MGLSourceQueryTests.m */; };
+		92F2C3ED1F0E3C3A00268EC0 /* MGLRendererFrontend.h in Headers */ = {isa = PBXBuildFile; fileRef = 92F2C3EC1F0E3C3A00268EC0 /* MGLRendererFrontend.h */; };
+		92F2C3EE1F0E3DC600268EC0 /* MGLRendererFrontend.h in Headers */ = {isa = PBXBuildFile; fileRef = 92F2C3EC1F0E3C3A00268EC0 /* MGLRendererFrontend.h */; };
```
looks like a duplicate of L:215

ivovandongen commented:
I ran the benchmark for the node bindings in `platform/node/test/benchmark.js`

branch | total time | time per request
--- | --- | ---
master | 237046.364ms | 47.41ms
||236751.378ms| 47.35ms
renderer-interface | 401433.224 | 80.28ms
||429934.557ms| 85.99ms

With the following settings:

setting  | value |
--- | --- |
mapPoolSize | 10 |
numRenderings | 5000
failurePercentage | 0
timeoutPercentage | 0
renderingTimeout | 5000
ratio | 2

jfirebaugh commented on src/mbgl/map/map.cpp:
```diff
@@ -727,44 +625,30 @@ void Map::updateAnnotation(AnnotationID id, const Annotation& annotation) {
 
 #pragma mark - Feature query api
 
 std::vector<Feature> Map::queryRenderedFeatures(const ScreenCoordinate& point, const RenderedQueryOptions& options) {
```
Can we do with `Map::query*Features` what we did for `Map::onLowMemory`, removing them from the `Map` interface and replacing calls to them with calls to methods on a concrete `RendererFrontend` classes?

ivovandongen replied:
We can, but it also facilitates `AnnotationIDs queryPointAnnotations(const ScreenBox&);`. Moving logic specific to Annotations into the renderer is maybe not that clean. It does clean up the different `RendererFrontend` implementations though..

ivovandongen replied:
@jfirebaugh I moved the feature querying to the renderer and used it directly on platforms where we are going to keep synchronous rendering and through the renderer frontend implementation for the others.

Still need to move the `map/query.hpp` header to `render/`

ivovandongen commented:
Ran the new [benchmarks](https://github.com/mapbox/mapbox-gl-native/issues/9449) for 50 iterations on my local machine (mac):

branch | test | mean (time) | mean (cpu) | std dev (time) | std dev (cpu)
--- | --- | --- | --- | --- | ---
**master** |||||
|| API_renderStill_reuse_map | 13691079 | 16040596 | 280928 | 764198
|| API_renderStill_reuse_map_switch_styles | 176082804 | 349192610 | 6938726 | 23015072
|| API_renderStill_recreate_map | 215923429 | 389453550 | 10762355 | 22513087
**renderer-interface** |||||
|| API_renderStill_reuse_map | 16050692 (**+17%**) | 20639946 | 611986 | 1338886
|| API_renderStill_reuse_map_switch_styles | 174545257 (-1%) | 336024850 | 6365576 | 22705308
|| API_renderStill_recreate_map | 211687565 (+2%) | 374861830 | 8861029 | 24739518

ivovandongen commented:
Seems like I was mistaken repetitions for iterations in google benchmark. Using more appropriate options, i come to different results:

`--benchmark_min_time=10 --benchmark_repetitions=4 --benchmark_filter=API_render*`

branch | test | mean (time) | mean (cpu) | iterations | iteration (time) | iteration (cpu)
--- | --- | --- | --- | --- | ---  | ---
**master** ||||||
|| API_renderStill_reuse_map | 5775795 | 3356289 | 3851 | 1599.82 | 871.54
|| API_renderStill_reuse_map_switch_styles | 116372011 | 278077468 | 47 | 2476000.23 | 5916541.87
|| API_renderStill_recreate_map | 170014910 | 321369112 | 40 | 4250372.75 | 8034227.80
**renderer-interface** ||||||
|| API_renderStill_reuse_map | 5765740 | 3351930 | 3823 | 1508.17 | 876.78
|| API_renderStill_reuse_map_switch_styles | 118266647 | 281591597 | 54 | 2190123.09 | 5214659.20
|| API_renderStill_recreate_map | 174577875 | 344321712 | 39 | 4476355.76 | 8828761.84

Seems to indicate that (re-)creating the map + the renderer frontend takes longer, but re-using is quite similar or better.

ivovandongen commented:
> Can the query* methods can be removed from Map now?

Yes. Seems to have gotten lost in one of the rebases

ivovandongen commented:
I've traced the slowness of the node bindings to `ImageManager::upload()` -> `Context::updateTexture()`. It's called the same amount of times, seemingly with the same arguments, but somehow executes slower on this branch than on master per call.

ivovandongen commented:
The slowdown in `ImageManager::upload()` was apparently caused by switching the order of `observer->onDidFinishRenderingMap()` (which calls the completion callback and subsequently reads the pixels from the view) and `painter->cleanup`. 

I inadvertently changed the order in Still rendering to match the order in Continuous rendering. We should confirm that the order in Continuous rendering is intentional.

ivovandongen commented:
@kkaefer You asked me to look into why the `test::sharedDisplay()` is not working as expected.

The problem lies in the fact that the destructor of `EGLDisplay` terminates the egl default display where the constructor initialises it. This would be fine if we use `test::sharedDisplay()` exclusively because then the `const shared_ptr` would only be cleared when all tests are over. Now, however, we use a mix of the shared display and using `EGLDisplay` directly through `HeadlessDisplay` and the display is terminated after some tests:

```
[ RUN      ] Map.LatLngBehavior
[New Thread 0x7fffed3b2700 (LWP 13042)]
[New Thread 0x7fffecbb1700 (LWP 13043)]
[New Thread 0x7fffe5ffb700 (LWP 13044)]
[New Thread 0x7fffdce2b700 (LWP 13045)]
[Thread 0x7fffed3b2700 (LWP 13042) exited]
[Thread 0x7fffdce2b700 (LWP 13045) exited]
[Thread 0x7fffe5ffb700 (LWP 13044) exited]
[Thread 0x7fffecbb1700 (LWP 13043) exited]

Thread 1 "mbgl-test" hit Breakpoint 1, mbgl::HeadlessDisplay::Impl::~Impl (this=0x30bc960, __in_chrg=<optimized out>) at ../../../platform/linux/src/headless_display_egl.cpp:45
45	    eglTerminate(display);
(gdb) bt
#0  mbgl::HeadlessDisplay::Impl::~Impl (this=0x30bc960, __in_chrg=<optimized out>) at ../../../platform/linux/src/headless_display_egl.cpp:45
#1  0x0000000001f6ede6 in std::default_delete<mbgl::HeadlessDisplay::Impl>::operator() (this=0x2afbc60, __ptr=0x30bc960) at /usr/include/c++/5/bits/unique_ptr.h:76
#2  0x0000000001f6ea75 in std::unique_ptr<mbgl::HeadlessDisplay::Impl, std::default_delete<mbgl::HeadlessDisplay::Impl> >::~unique_ptr (this=0x2afbc60, __in_chrg=<optimized out>) at /usr/include/c++/5/bits/unique_ptr.h:236
#3  0x0000000001f6e8fa in mbgl::HeadlessDisplay::~HeadlessDisplay (this=0x2afbc60, __in_chrg=<optimized out>) at ../../../platform/linux/src/headless_display_egl.cpp:62
#4  0x00000000020768e6 in std::_Sp_counted_ptr<mbgl::HeadlessDisplay*, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x2b30560) at /usr/include/c++/5/bits/shared_ptr_base.h:374
#5  0x00000000017f170c in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x2b30560) at /usr/include/c++/5/bits/shared_ptr_base.h:150
#6  0x00000000017f01cf in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffffffd5a0, __in_chrg=<optimized out>) at /usr/include/c++/5/bits/shared_ptr_base.h:659
#7  0x0000000001843d74 in std::__shared_ptr<mbgl::HeadlessDisplay, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffffffd598, __in_chrg=<optimized out>) at /usr/include/c++/5/bits/shared_ptr_base.h:925
#8  0x0000000001843d90 in std::shared_ptr<mbgl::HeadlessDisplay>::~shared_ptr (this=0x7fffffffd598, __in_chrg=<optimized out>) at /usr/include/c++/5/bits/shared_ptr.h:93
#9  0x0000000001fc6e02 in mbgl::HeadlessBackend::~HeadlessBackend (this=0x7fffffffd580, __in_chrg=<optimized out>) at ../../../platform/default/mbgl/gl/headless_backend.cpp:18
#10 0x00000000018a0cfc in MapTest<mbgl::StubFileSource>::~MapTest (this=0x7fffffffd4f0, __in_chrg=<optimized out>) at ../../../test/map/map.test.cpp:63
#11 0x00000000018926e5 in Map_LatLngBehavior_Test::TestBody (this=0x2afb650) at ../../../test/map/map.test.cpp:103
#12 0x00000000020a682a in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#13 0x000000000208c041 in testing::Test::Run() ()
#14 0x000000000208d020 in testing::TestInfo::Run() ()
#15 0x000000000208d783 in testing::TestCase::Run() ()
#16 0x0000000002095909 in testing::internal::UnitTestImpl::RunAllTests() ()
#17 0x00000000020a7168 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#18 0x000000000209538f in testing::UnitTest::Run() ()
#19 0x0000000001907e4e in RUN_ALL_TESTS () at ../../../mason_packages/linux-x86_64/gtest/1.8.0-cxx11abi/include/gtest/gtest.h:2233
#20 0x0000000001907da3 in mbgl::runTests (argc=1, argv=0x7fffffffdca8) at ../../../test/src/mbgl/test/test.cpp:14
#21 0x0000000001c08efd in main (argc=1, argv=0x7fffffffdca8) at ../../../platform/default/mbgl/test/main.cpp:19

```

In a follow-up test using `test::sharedDisplay` `eglInitialise()` is not called as the shared_ptr is still valid, but the EGL display is not.

ivovandongen commented:
@kkaefer As discussed offline I moved the display sharing into a factory function on `HeadlessDisplay`: https://github.com/mapbox/mapbox-gl-native/pull/9327/commits/9a930e246fe1df1543b4a999a026c734377ea259

ivovandongen commented:
@jfirebaugh @kkaefer 
  Changes in the way the backend scope is handled led to recurrence of issues on Android like:

```
eglMakeCurrent() returned error 12301
eglMakeCurrent error 300d (EGL_BAD_SURFACE)
```

These errors were caused by the way we deal with nested BackendScopes icm with the manual ref counting on NativeMapView on Android. I’ve removed the manual ref counting on Android and added a check on BackendScope to only activate a backend once and only deactivate if that scope actually activated the backend. Tests should hopefully pass now.  

I’d love some eyes on the tests though. Specifically, is it a correct assumption that a currently active backend should be deactivated before activating a new backend or is activation of a backend implicitly a deactivation of any active backend at that time?

Quite possibly related: 
- https://github.com/mapbox/mapbox-gl-native/issues/9458 
- https://github.com/mapbox/mapbox-gl-native/issues/6647  

ivovandongen commented:
@kkaefer Noticed that animations are choppier in this branch than on master. Seems like fewer frames are rendered than before.

ivovandongen commented:
@frederoni It seems there is a slight difference between macOS and iOS when calling [nativeView setNeedsDisplay]. On macOS, when doing this while renderSync is in progress, it is not queued behind the current redraw, whereas it is on iOS. To make sure that animations work properly on both platforms I had to make sure it works with an `AsyncTask` on macOS where it is not needed on iOS.

I kept a single `RendererFrontend` implementation, with a switch to enable asynchronous behaviour in https://github.com/mapbox/mapbox-gl-native/pull/9327/commits/ce3538d30016de4c6d3839eee1e368b1c27b38f6. I rather not make two separate implementations. Let me know how this looks to you.

@kkaefer A similar problem existed on glfw where the dirty flag was cleared too late so another iteration could not be triggered from anination frame callbacks. I addressed this in https://github.com/mapbox/mapbox-gl-native/pull/9327/commits/f290c6ddfb5ad0358a6c7ba38ab12420f4441ad8
